### ISSUE TITLE
Does not work after recent Agents of Sabotage GTAO update.

### ISSUE THREAD

**Issue by ElectroBytezLV** [JU5TDIE/Lester-Ver2.0#11]
Script simply does not work after the recent Agents of Sabotage DLC update. When you press the respective key, it does not do anything in game. You can see the script actually doing stuff in the console, but the actual keypresses are not being reflected in game.

Rockstar changed something in the most recent update relating to macros, hotkeys and keypresses that are from an external program, because I had the same issue with my AHK script and only started working after a few changes to the actual hotkeys.

I talked to a person that knows python well and they said that they got it to work after they used pynput directly to emulate/send the keypresses instead of the keyboard library, but I am unsure about the technical details.

**Comment by JU5TDIE** [comment]
Sorry for the late reply, I've been quite busy. Anyways, I was already aware of this issue but I couldn't find the clear solution yet. The key inputs work most of the time but sometimes they do not.

And about the keyboard library issue, I have already tried with pynput library for key inputs, but it didn't even work before the Agents of Sabotage DLC. After testing with some libraries, ctypes was the only library that worked flawlessly with the direct input key values. However, I'm not sure if this is reliable for the main branch so I'll create a fork of this later. Feel free to test it and let me know if you encounter any issues.

**Comment by ElectroBytezLV** [comment]
So a little information, I'm not a good programmer so I also tried converting the casino fingerprint script to use ctypes or pynput with chatgpt, but that didnt change anything for me (it did function just not in GTA).
I tried MANY different combos and methods to try and make this work, for example, installing the recommended python version and the libraries with the exact version, using the latest python version with the latest versions of each library, etc.
Tried launching the script in multiple ways: with admin, from a cmd window, via the main.py, via the exe, putting it in the same folder as GTA 5, etc. Launching the script before or after launching GTA 5. Nothing seemed to solve it.

Then magically I somehow managed to get it working. I have no clue how that could be a thing, but it seems like the script only works for me IF I launch GTA 5 from rockstar launcher OR have rockstar launcher open in the background (NOT minimized to the system tray).
If I minimize R* launcher or close it and launch the game from epic games or from my desktop shortcut, the script does not work. If I launch the game from R* launcher or launch the game from the shortcut or epic games when R* launcher is maximized, then it seemed to also work. I will do more testing related to these, but thats one thing that was fairly consistent in my testing.

It is quite inconsistent though, because one person tried it and it worked without any issues for them not matter how they launch their game or what they do.

**Comment by JU5TDIE** [comment]
Well, I don't know how reliable chatgpt is nowadays, but I doubt it gave you the correct answers. From my experience, they were wrong most of the time when it came to resolving some complex problems.

> Then magically I somehow managed to get it working. I have no clue how that could be a thing, but it seems like the script only works for me IF I launch GTA 5 from rockstar launcher OR have rockstar launcher open in the background (NOT minimized to the system tray).

I bought gta on rockstar launcher so I always launch gta from rockstar launcher but the script does not work even for me sometimes. I also tested multiple times with different setup, but it's too random that I couldn't find the pattern unfortunately.

Here is [the branch](https://github.com/JU5TDIE/Lester-Ver2.0/tree/ctypes-fix) I updated. As I explained above, it uses ctypes library with direct input, which means it **might** conflict with battleye. Test it at your own risk and let me know if you encounter any issues.

**Comment by ElectroBytezLV** [comment]
The answers chatgpt gave seemed to look and work fine, just as I mentioned they didnt change anything in game. Either way I also havent been able to find a pattern of when it works or doesnt work. 
I will try out the new branch later on though to see if it works, thank you for the response and help trying to solve the issue.

**Comment by ElectroBytezLV** [comment]
I have tried the updated branch and it seems to work perfectly now first time I opened it. If it worked perfect for you and me, I guess you can push this to the main branch and add a disclaimer about battleye.
Thank you for the help of solving this issue.

I wanted to ask another question. I've noticed when using this previously that quite often the script (doesnt matter if exe version or py version) closes by itself even if GTA is still open. What could be the issue or how could you solve it? Maybe a temporary solution is to launch the py script from a terminal window and it wont close that automatically? I'm not sure of the pattern on when or why it closes but I just have noticed often that it disappears from my taskbar and I have to reopen it.

Edit: I also tried it on cayo and the cayo retro fingerprint is quite unreliable - it often just spams random buttons without actually resolving the puzzle correctly (probably have to increase the delay between presses, but I havent tried it yet)

**Comment by JU5TDIE** [comment]
Could you test it again? Because I tried to test cayo fingerprint hack with the ctypes-fix branch but it suddenly stopped working for me sadly. And I don't think it's the problem with specific library.

This is how I tested with both main and ctypes-fix branch. I started the program as soon as my character was loaded. The key inputs were unresponsive as usual, however the entire program went unresponsive after 8~10 minutes this time ([pynput keyboard event](https://github.com/JU5TDIE/Lester-Ver2.0/blob/4f014a13ddff32f84a9fe2e8adeec501318f94de/main.py#L64) not responding and I had to restart every time). After restarting the program, it worked flawlessly.

Strangely, it worked in the first place when I put 0.1 millisecond (0.0001 second) delay [between PressKey and ReleaseKey function.](https://github.com/JU5TDIE/Lester-Ver2.0/blob/2bfbc24f4cb4f394a75d979851d0f94a25561411/directinput.py#L68) But this also went unresponsive after 8~10 minutes and had to restart the program. (I assume battleye is doing something under the hood. The program never went unresponsive when I started gta with battleye off)

I'm extremely sure I found the pattern of this annoying issue, but I still haven't found the solution to fix this. I'll let you know when I push the fix.

> I've noticed when using this previously that quite often the script (doesnt matter if exe version or py version) closes by itself even if GTA is still open. What could be the issue or how could you solve it?

Did you press F4 by any chance? It never closed by itself on my PC. Also, if you run it from the source on terminal, there should be an error output. Feel free to create a new issue and don't forget to post that error output together ;)

> Edit: I also tried it on cayo and the cayo retro fingerprint is quite unreliable - it often just spams random buttons without actually resolving the puzzle correctly

That's probably because the script is getting each key's hex value from the dictionary every time, which consumes a little bit of time (sorry for the non-optimized code design, I had to make a quick fix). This only applies to ctypes-fix branch so you may also try with the main branch since they both have the same issue on my PC

**Comment by ElectroBytezLV** [comment]
TLDR;

For some reason, main branch is working for me now (as you can see from my previous messages it did not work at all).
I did not experience unresponsiveness for either main or ctypes fix branch.
Casino fingerprint works perfectly now for both branches.
Cayo retro fingerprint does not work correctly for either branches.



Ok so I did some more testing. To my surprise the main branch somehow is working for me now in game (its actually solving stuff) where as you can see from my previous messages, it did not work at all. I dont understand how or why the main branch is working for me now.

I was probably testing both branches for more than 10-30 minutes and I did not experience any of them freezing or crashing and me having to reopen then.

Also yes I tried and saw that F4 closes the script which I found quite odd, since its a common button that you could accidentally press or press it for other purposes and the script would close. In my opinion - remove it so the only way to close the script is to manually close the window, the exe or CTRL+B in the terminal window. So it doesnt close on accident.
Not sure if I just got unlucky but I didnt experience it closing randomly (so maybe I was pressing F4 often even though I dont remember using it often).

Now I tried both main and ctypes branch on cayo retro fingerprint and I can report that is basically did not work almost every time. It just spams random buttons and does not solve the fingerprint correctly almost everytime. Which is weird because it used to work perfectly before battleye.
I also experimented with casino fingerprint and keypad and both ctypes and main branch work PERFECTLY for casino fingerprint hack. 
But for some reason sometimes both branches fail on the keypad even though I pressed F6 on the last pattern correctly (it seemed to fail more often on the main branch and cytpes branch only failed one of the hacks out of like 10 tries). No one does the keypad hack anyways but obviously it would be nice if it was working correctly.

**Comment by dananon929** [comment]
> Could you test it again? Because I tried to test cayo fingerprint hack with the ctypes-fix branch but it suddenly stopped working for me sadly. And I don't think it's the problem with specific library.
> 

I noticed an issue before the DLC dropped and made some changes with the hack py files. Instead of using
```
keyboard.press_and_release(key)
time.sleep(0.025)
```
I had used
```
keyboard.press(key)
time.sleep(0.05)
keyboard.release(key)
time.sleep(0.05)
```
since I realized some update required a delay between press and release. My version of Lester-VER2.0 is still working.

**Comment by ElectroBytezLV** [comment]
> I had used
> 
> ```
> keyboard.press(key)
> time.sleep(0.05)
> keyboard.release(key)
> time.sleep(0.05)
> ```

Can confirm this is working correctly when changing in the main branch. So I would assume a similar behavior should be adapted for the ctypes fix as well for more reliable operation compared to main branch.

**Comment by dananon929** [comment]
> Can confirm this is working correctly when changing in the main branch

Glad it works! 😄

**Comment by JU5TDIE** [comment]
> I did not experience unresponsiveness for either main or ctypes fix branch.
https://github.com/JU5TDIE/Lester-Ver2.0/issues/11#issuecomment-2573708904

Well.. this is the closest thing I could find about battleye issue. Unfortunately, I couldn't find what causes this :(

> `C:\Users\USER\Documents\Rockstar Games\Launcher\launcher.log`
[WARNING] [21312] [gamelaunch] 300 seconds elapsed and still no game window.
[WARNING] [21312] [gamelaunch] No game or BattlEye window exists for game tracking.
[WARNING] [21312] [gamelaunch] 300 seconds elapsed and still no game window.

I believe that battleye is blocking the key inputs when it's continuously searching for the game window. The delay is only required when it's searching for the game window and after 600 seconds, the program and its key inputs work perfectly fine without any delay between press and release function (have to restart the program though). The timer starts when the game starts so if this issue is reproducible from other computers, I can maybe update the program to read the log and adjust the delay by itself. 

> I had used
> 
> ```
> keyboard.press(key)
> time.sleep(0.05)
> keyboard.release(key)
> time.sleep(0.05)
> ```

Or I can simply implement this to the main branch right away but after 600 seconds this is not required at all. If there isn't any other way to resolve this issue besides this temporary fix, I'll consider implementing this with the reduced delay. (sleeping 0.1 seconds for sending a single key seems too slow for me)

Since that warning log could be my issue, please let me know if you also find this particular warning log from launcher.log. Regardless of this battleye warning log, I can play gta online without any issues though.

> Cayo retro fingerprint does not work correctly for either branches.

It worked perfectly fine with main branch but not well with ctypes-fix branch for me. I strongly recommend sticking to the main branch by the way since ctypes-fix branch does not fix this issue anymore and they are almost identical. 

> Also yes I tried and saw that F4 closes the script which I found quite odd, since its a common button that you could accidentally press or press it for other purposes and the script would close. In my opinion - remove it so the only way to close the script is to manually close the window, the exe or CTRL+B in the terminal window. So it doesnt close on accident.

I'll add a config file for key bindings in the next update then. If you are running the program from source, you can simply remove the shutdown key by removing [this line](https://github.com/JU5TDIE/Lester-Ver2.0/blob/4f014a13ddff32f84a9fe2e8adeec501318f94de/main.py#L65) actually. Or key combination like [this](https://github.com/boppreh/keyboard?tab=readme-ov-file#keyboardsendhotkey-do_presstrue-do_releasetrue).

**Comment by Jacksonidis** [comment]
hello guys can you please help me with the keypad cracker doesnt work i press f6 when the latest pattern blink but nothing happens
 what can i change to make it work ? https://imgur.com/a/4FgwH63

**Comment by JU5TDIE** [comment]
> hello guys can you please help me with the keypad cracker doesnt work i press f6 when the latest pattern blink but nothing happens what can i change to make it work ? https://imgur.com/a/4FgwH63


```python
if np.mean(crop_img) == 0:
    keyboard.press(key)
    time.sleep(0.05)
    keyboard.release(key)
    time.sleep(0.05)

    time.sleep(0.025)
```
```python
for key in keyboardgo:
    keyboard.press(key)
    time.sleep(0.05)
    keyboard.release(key)
    time.sleep(0.05)
    if key == 's' or 'w':
        time.sleep(0.025)
    if key == 'return':
        time.sleep(1.95)
```
This should work.

**Comment by dawackyman** [comment]
So what is the fix to get the hacks working again? I read through this thread and am not understanding clearly. Can you explain it simply to me?

**Comment by JU5TDIE** [comment]
> So what is the fix to get the hacks working again? I read through this thread and am not understanding clearly. Can you explain it simply to me?

The hacks work perfectly fine actually. It's just the battleye blocking the key inputs for the first 10 minutes. These fixes are not required at all after 10 minutes of waiting in any gta online session.

**Comment by dawackyman** [comment]
> > So what is the fix to get the hacks working again? I read through this thread and am not understanding clearly. Can you explain it simply to me?
> 
> The hacks work perfectly fine actually. It's just the battleye blocking the key inputs for the first 10 minutes. These fixes are not required at all after 10 minutes of waiting in any gta online session.

I'll give that a go then.. Before this wasn't the case for me. So the release version works exactly as expected? Or should I download the source and run it from there?

**Comment by dawackyman** [comment]
> > So what is the fix to get the hacks working again? I read through this thread and am not understanding clearly. Can you explain it simply to me?
> 
> The hacks work perfectly fine actually. It's just the battleye blocking the key inputs for the first 10 minutes. These fixes are not required at all after 10 minutes of waiting in any gta online session.

Can confirm that even in the newest release version this does not work even after waiting 10 minutes. At least for me.

**Comment by Jacksonidis** [comment]
https://imgur.com/a/tRMh8uJ

**Comment by JU5TDIE** [comment]
> Can confirm that even in the newest release version this does not work even after waiting 10 minutes. At least for me.

`C:\Users\<username>\Documents\Rockstar Games\Launcher\launcher.log` can you open this log file and check if there are two "300 seconds elapsed and still no game window." logs? If you see only one or none of them, it means you didn't wait for 10 minutes.

**Comment by JU5TDIE** [comment]
> https://imgur.com/a/tRMh8uJ

Please create a new issue if it's irrelevant to this issue.
